File: src/data_structure/linked_list/remove_duplicate.cpp
/*
  Remove all duplicate elements from a sorted linked list
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/
Node* RemoveDuplicates(Node *head)
{
    if (head == NULL || head->next == NULL) return head;

    Node *pre, *cur, *temp;
    pre = head;
    cur = head->next;
    while (cur != NULL)
    {
        if (pre->data == cur->data)
        {
            temp = cur->next;
            pre->next = cur->next;
            free(cur);
            cur = temp;
        }
        else
        {
            pre = cur;
            cur = cur->next;
        }
    }
    return head;
}

File: src/data_structure/linked_list/remove_duplicate.java
Node RemoveDuplicates(Node head) {
    if (head == null || head.next == null) {
        return head;
    }
    Node pre, cur, temp;
    pre = head;
    cur = head.next;
    while (cur != null) {
        if (pre.data == cur.data) {
            temp = cur.next;
            pre.next = temp;
            cur = temp;
        } else {
            pre = cur;
            cur = cur.next;
        }
    }
    return head;
}

